src/Stratis.Bitcoin.Features.SmartContracts/Consensus/Rules/CanGetSenderRule.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using NBitcoin;
+using Stratis.Bitcoin.Consensus;
+using Stratis.Bitcoin.Consensus.Rules;
+using Stratis.Bitcoin.Features.Consensus.CoinViews;
+using Stratis.Bitcoin.Features.Consensus.Rules;
+using Stratis.Bitcoin.Features.MemoryPool;
+using Stratis.SmartContracts.Core;
+using Stratis.SmartContracts.Core.Util;
+using Block = NBitcoin.Block;
+
+namespace Stratis.Bitcoin.Features.SmartContracts.Consensus.Rules
+{
+    public class CanGetSenderRule : UtxoStoreConsensusRule, ISmartContractMempoolRule
+    {
+        private readonly ISenderRetriever senderRetriever;
+
+        public CanGetSenderRule(ISenderRetriever senderRetriever)
+        {
+            this.senderRetriever = senderRetriever;
+        }
+
+        public override Task RunAsync(RuleContext context)
+        {
+            Block block = context.ValidationContext.BlockToValidate;
+            IList<Transaction> processedTxs = new List<Transaction>();
+
+            foreach (Transaction transaction in block.Transactions)
+            {
+                CheckTransactionInsideBlock(transaction, this.PowParent.UtxoSet, processedTxs);
+                processedTxs.Add(transaction);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private void CheckTransactionInsideBlock(Transaction transaction, ICoinView coinView, IList<Transaction> blockTxs)
+        {
+            // If wanting to execute a contract, we must be able to get the sender.
+            if (transaction.Outputs.Any(x => x.ScriptPubKey.IsSmartContractExec()))
```

Comment:
Could reuse this logic as it's the same as below.

Reply:
Thanks you are right but the GetSender call is different so it unfortunately isn't  a clean methodisation. Could  look at it in future though